src/target/riscv/riscv.c

```diff
@@ -1199,7 +1199,7 @@ static int riscv_add_breakpoint(struct target *target, struct breakpoint *breakp
 	assert(breakpoint);
 	if (breakpoint->type == BKPT_SOFT) {
 		/** @todo check RVC for size/alignment */
-		if (!(breakpoint->length == 4 || breakpoint->length == 2)) {
+		if (!(breakpoint->length == 4 || (breakpoint->length == 2 && riscv_supports_extension(target, 'C')))) {
 			LOG_TARGET_ERROR(target, "Invalid breakpoint length %d", breakpoint->length);
```

Comment:
Please, consider improving also the error message. Thanks.

Example:

```suggestion
			LOG_TARGET_ERROR(target, "Invalid breakpoint length %d (supported lengths: %s)", breakpoint->length, riscv_supports_extension(target, 'C') ? "2, 4" : "4");
```

Reply:
Done